test/integration/network_extension_discovery_integration_test.cc

```diff
@@ -0,0 +1,707 @@
+  size_t index = 0;
+  for (const Json::ObjectSharedPtr& obj_ptr : json->getObjectArray("configs")) {
+    EXPECT_TRUE(expected_types[index].compare(obj_ptr->getString("@type")) == 0);
+    index++;
+  }
+
+  envoy::admin::v3::ConfigDump config_dump;
+  TestUtility::loadFromJson(response->body(), config_dump);
+  EXPECT_EQ(5, config_dump.configs_size());
+  envoy::admin::v3::EcdsConfigDump ecds_config_dump;
+  config_dump.configs(2).UnpackTo(&ecds_config_dump);
+  envoy::config::core::v3::TypedExtensionConfig filter_config;
+  test::integration::filters::TestDrainerNetworkFilterConfig network_filter_config;
+  // Verify the first filter.
+  EXPECT_EQ("1", ecds_config_dump.ecds_filters(0).version_info());
+  EXPECT_TRUE(ecds_config_dump.ecds_filters(0).ecds_filter().UnpackTo(&filter_config));
+  filter_config.typed_config().UnpackTo(&network_filter_config);
+  EXPECT_TRUE(verifyConfigDumpData(filter_config, network_filter_config));
+  // Verify the second filter.
+  EXPECT_EQ("1", ecds_config_dump.ecds_filters(1).version_info());
+  EXPECT_TRUE(ecds_config_dump.ecds_filters(1).ecds_filter().UnpackTo(&filter_config));
+  filter_config.typed_config().UnpackTo(&network_filter_config);
+  EXPECT_TRUE(verifyConfigDumpData(filter_config, network_filter_config));
+}
+
+// ECDS config dump test with specified resource and regex name search.
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, TwoSubscriptionsConfigDumpWithResourceAndRegex) {
+  DISABLE_IF_ADMIN_DISABLED; // Uses admin interface.
+  two_connections_ = true;
+  on_server_init_function_ = [&]() { waitXdsStream(); };
+  addFilterChain();
+  addDynamicFilter("foo", true);
+  addDynamicFilter("bar", false, true, false, true);
+  initialize();
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+
+  sendXdsResponse("foo", "1", 3);
+  sendXdsResponse("bar", "1", 4, false, true);
+  test_server_->waitForCounterGe("extension_config_discovery.network_filter.foo.config_reload", 1);
+  test_server_->waitForCounterGe("extension_config_discovery.network_filter.bar.config_reload", 1);
+  BufferingStreamDecoderPtr response;
+  EXPECT_EQ("200",
+            request("admin", "GET", "/config_dump?resource=ecds_filters&name_regex=.a.", response));
+
+  envoy::admin::v3::ConfigDump config_dump;
+  TestUtility::loadFromJson(response->body(), config_dump);
+  EXPECT_EQ(1, config_dump.configs_size());
+  envoy::admin::v3::EcdsConfigDump::EcdsFilterConfig ecds_msg;
+  config_dump.configs(0).UnpackTo(&ecds_msg);
+  EXPECT_EQ("1", ecds_msg.version_info());
+  envoy::config::core::v3::TypedExtensionConfig filter_config;
+  EXPECT_TRUE(ecds_msg.ecds_filter().UnpackTo(&filter_config));
+  EXPECT_EQ("bar", filter_config.name());
+  test::integration::filters::TestDrainerNetworkFilterConfig network_filter_config;
+  filter_config.typed_config().UnpackTo(&network_filter_config);
+  EXPECT_EQ(4, network_filter_config.bytes_to_drain());
+}
+
+} // namespace
```

Comment:
2 high-level questions:
1. Are there planned integration tests that test the interaction between listener updates (LDS) and network-filter updates (ECDS)?
2. Can the HCM be updated? If so I think more tests are needed in that space because of the lifetime of some objects.
3. Please add tests that have ongoing connections during an ECDS update and show that the "old" config is still used for these filters.

Reply:
I added LDS integration and a test that verifies that old config is used for connections that were established before ECDS update. Due to (3), I believe the behavior here would be the same, connections will get the HCM config that is updated at the time of connection establishment. Did you have another behavior in mind which I missed?